_posts/2023-09-07-develop-functions-locally.md

```diff
@@ -0,0 +1,343 @@
+
+Run the echo function:
+
+```bash
+faas-cli local-run echo
+```
+
+Invoke the function with in invalid api-key:
+
+```bash
+$ curl -s http://127.0.0.1:8080 \
+  -H "Authorization: Bearer invalid"
+  -H "Content-Type: text/plain" \
+  -d "Greetings from OpenFaaS"
+
+HTTP/1.1 401 Unauthorized
+Connection: keep-alive
+Content-Length: 15
+Content-Type: text/html; charset=utf-8
+Date: Thu, 07 Sep 2023 11:20:50 GMT
+Etag: W/"f-DQI+tV0HvT0NM/9khJk6PQXU+K4"
+Keep-Alive: timeout=5
+X-Duration-Seconds: 0.003185
+
+Invalid api key%
+```
+
+Invoking the function with the correct api-key that we saved in the secret should return a HTTP 200 status code.
+
+```bash
+$ curl -s http://127.0.0.1:8080 \
+  -H "Authorization: Bearer secret-access-token"
+  -H "Content-Type: text/plain" \
+  -d "Greetings from OpenFaaS"
+
+HTTP/1.1 200 OK
+Connection: keep-alive
+Content-Length: 66
+Content-Type: text/html; charset=utf-8
+Date: Thu, 07 Sep 2023 11:15:30 GMT
+Etag: W/"42-ql9Q2lRIiMOJhodGVWvMjGR/Ezw"
+Keep-Alive: timeout=5
+X-Duration-Seconds: 0.004418
+
+{"body":"\"Greetings from OpenFaaS\"","content-type":"text/plain"}%
+```
+
+## Automatic rebuilds by watching the filesystem
+
+Another feature that can be helpful to iterate on functions quickly during development is the built in watch functionality of the CLI.
+
+The `--watch` flag can be used with both the `local-run` and `up` command. Adding the flag will tell the cli to watch the filesystem for any changes to the function source files and automatically re-build and deploy functions on save.
+
+When using `--watch` flag with `faas-cli up` it is recommended to also set `--tag=digest`. This ensures unique image tags are generated for each build. The next section goes into more detail about the `--tag` flag.
+
+## Generated image tags
+
+All OpenFaaS functions are built into container images. By default if no image tag is included for a function in the stack.yml file the `:latest` tag is used. When iterating over functions and pushing them to an image registry it is a best practice to organise different image versions using tags instead of always pushing to `:latest`.
+
+There are two options to set tags for function images.
+
+1. Let the faas-cli generate the tag automatically.
+2. Set the image tag in the stack.yml file.
+
+The `--tag` option can be used with the `build`, `push` and `deploy` sub-commands of the faas-cli. If this flag is provided, image tags for functions will automatically be generated based on available metadata. This can be either Git metadata like the commit sha or branch name or digest of the function handler content.
+
+The generated tag is always suffixed to any tag defined in the stack.yml file or `latest` if no tag is defined.
+
+Some examples:
+
+When using the flag `--tag=sha` the image tag used in the stack.yml file is suffixed with the short Git SHA. e.g
+
+```yaml
+functions:
+  echo:
+    lang: node18
+    handler: ./echo
+    image: echo:0.2
```

Comment:
People will run this and then wonder why it doesn't work.

We should have a registry as a prefix, even if it's ttl.sh/openfaas/

Reply:
@welteki you've updated the wrong field, the handler is the image?